VideoConversionLifecycleManager & VncRecorderLifecycleManager should not fail the suite

##### Issue Overview

VideoConversionLifecycleManager & VncRecorderLifecycleManager both use `@Observes` for lifecycle.

As `@Observes` events are handled directly in the event chain then extra care (try/catch) needs to be taken on event methods that may fail. Other observers must still be called even on an error, else cleanups are not run for example.

Generally, all `@Observes` methods should be safe to call (try/catch), with their own error handling, else failures break the event chain.

##### Expected Behaviour

At least an error message on failure, but not the failure of the test suite.

##### Current Behaviour

Any errors in `@Observes` methods cause the test suite to fail, eg:

`
org.arquillian.cube.spi.CubeControlException: Could not create flv2mp4
	at org.arquillian.cube.spi.CubeControlException.failedCreate(CubeControlException.java:15)
	at org.arquillian.cube.docker.impl.model.DockerCube.create(DockerCube.java:115)
	at org.arquillian.cube.docker.drone.VideoConversionLifecycleManager.startConversion(VideoConversionLifec

.....

Caused by: com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 192.168.65.1:53: no such host"}
`

The above example is due to the fact the image is blocked by policy, but it should not break the suite.

--- a/docker/drone/src/main/java/org/arquillian/cube/docker/drone/CubeDroneConfiguration.java
+++ b/docker/drone/src/main/java/org/arquillian/cube/docker/drone/CubeDroneConfiguration.java
@@ -36,6 +36,8 @@ public class CubeDroneConfiguration {
      */
     private String containerNamePrefix = null;
 
+    private String dockerRegistry = "";
+
     public static CubeDroneConfiguration fromMap(Map<String, String> config) {
         CubeDroneConfiguration cubeDroneConfiguration = new CubeDroneConfiguration();
 
@@ -61,6 +63,21 @@ public class CubeDroneConfiguration {
 
         if (config.containsKey("containerNamePrefix")) {
             cubeDroneConfiguration.containerNamePrefix = config.get("containerNamePrefix");
+        }
+
+        if (config.containsKey("dockerRegistry")) {
+            String dockerRegistry = config.get("dockerRegistry");
+
+            if(null != dockerRegistry) {
+
+                dockerRegistry = dockerRegistry.trim();
+
+                if(!dockerRegistry.endsWith("/")){
+                    dockerRegistry = dockerRegistry + "/";
+                }
+
+                cubeDroneConfiguration.dockerRegistry = dockerRegistry;
+            }
         }
 
         return cubeDroneConfiguration;
@@ -109,7 +126,11 @@ public class CubeDroneConfiguration {
     public String getContainerNamePrefix(){
         return this.containerNamePrefix;
     }
-    
+
+    public String getDockerRegistry() {
+        return dockerRegistry;
+    }
+
     public static enum RecordMode {
         ALL, ONLY_FAILING, NONE;
     }

--- a/docker/drone/src/main/java/org/arquillian/cube/docker/drone/SeleniumContainers.java
+++ b/docker/drone/src/main/java/org/arquillian/cube/docker/drone/SeleniumContainers.java
@@ -51,6 +51,7 @@ public class SeleniumContainers {
     private final String vncContainerName;
     private final String conversionContainerName;
     private final int seleniumBoundedPort;
+    private final String dockerRegistry;
 
     private SeleniumContainers(String browser, CubeDroneConfiguration cubeDroneConfiguration) {
         this.browser = browser;
@@ -77,24 +78,23 @@ public class SeleniumContainers {
                 this.seleniumBoundedPort = 14444;
                 break;
         }
-        
+
+        this.dockerRegistry = cubeDroneConfiguration.getDockerRegistry();
         this.videoRecordingFolder = VolumeCreator.createTemporaryVolume(DEFAULT_PASSWORD);
-        
-        this.seleniumContainer = createSeleniumContainer(browser, cubeDroneConfiguration, this.seleniumBoundedPort);
-        this.vncContainer = createVncContainer(this.videoRecordingFolder, this.seleniumContainerName);
+        this.seleniumContainer = createSeleniumContainer(browser, cubeDroneConfiguration, this.seleniumBoundedPort, this.dockerRegistry);
+        this.vncContainer = createVncContainer(this.videoRecordingFolder, this.seleniumContainerName, this.dockerRegistry);
         final Path targetVolume = VideoFileDestination.resolveTargetDirectory(cubeDroneConfiguration);
-        this.videoConverterContainer = createVideoConverterContainer(targetVolume);
-        
+        this.videoConverterContainer = createVideoConverterContainer(targetVolume, this.dockerRegistry);
     }
 
     public static SeleniumContainers create(String browser, CubeDroneConfiguration cubeDroneConfiguration) {
         return new SeleniumContainers(browser, cubeDroneConfiguration);
     }
 
-    private static CubeContainer createVideoConverterContainer(Path dockerVolume) {
+    private static CubeContainer createVideoConverterContainer(Path dockerVolume, String dockerRegistry) {
 
         CubeContainer cubeContainer = new CubeContainer();
-        cubeContainer.setImage(Image.valueOf(CONVERSION_IMAGE));
+        cubeContainer.setImage(Image.valueOf(dockerRegistry + CONVERSION_IMAGE));
 
         cubeContainer.setBinds(
             Arrays.asList(convertToBind(dockerVolume, VOLUME_DIR, "Z"))
@@ -114,10 +114,10 @@ public class SeleniumContainers {
         return cubeContainer;
     }
 
-    private static CubeContainer createVncContainer(final Path dockerVolume, String seleniumContainerName) {
+    private static CubeContainer createVncContainer(final Path dockerVolume, String seleniumContainerName, String dockerRegistry) {
 
         CubeContainer cubeContainer = new CubeContainer();
-        cubeContainer.setImage(Image.valueOf(VNC_IMAGE));
+        cubeContainer.setImage(Image.valueOf(dockerRegistry + VNC_IMAGE));
 
         cubeContainer.setBinds(
             Arrays.asList(convertToBind(dockerVolume, VOLUME_DIR, "Z"))
@@ -175,7 +175,7 @@ public class SeleniumContainers {
         }
     }
 
-    private static CubeContainer createSeleniumContainer(String browser, CubeDroneConfiguration cubeDroneConfiguration, int seleniumBoundedPort) {
+    private static CubeContainer createSeleniumContainer(String browser, CubeDroneConfiguration cubeDroneConfiguration, int seleniumBoundedPort, String dockerRegistry) {
 
         if (cubeDroneConfiguration.isBrowserDockerfileDirectorySet()) {
             return createCube(cubeDroneConfiguration.getBrowserDockerfileLocation(), seleniumBoundedPort);
@@ -183,19 +183,19 @@ public class SeleniumContainers {
             if (cubeDroneConfiguration.isBrowserImageSet()) {
                 return configureCube(cubeDroneConfiguration.getBrowserImage(), seleniumBoundedPort);
             } else {
-                return useOfficialSeleniumImages(browser, seleniumBoundedPort);
+                return useOfficialSeleniumImages(browser, seleniumBoundedPort, dockerRegistry);
             }
         }
     }
 
-    private static CubeContainer useOfficialSeleniumImages(String browser, int seleniumBoundedPort) {
+    private static CubeContainer useOfficialSeleniumImages(String browser, int seleniumBoundedPort, String dockerRegistry) {
         String version = SeleniumVersionExtractor.fromClassPath();
 
         switch (browser) {
             case "firefox":
-                return configureCube(String.format(FIREFOX_IMAGE, version), seleniumBoundedPort);
+                return configureCube(dockerRegistry + String.format(FIREFOX_IMAGE, version), seleniumBoundedPort);
             case "chrome":
-                return configureCube(String.format(CHROME_IMAGE, version), seleniumBoundedPort);
+                return configureCube(dockerRegistry + String.format(CHROME_IMAGE, version), seleniumBoundedPort);
             default:
                 throw new UnsupportedOperationException(
                     "Unsupported browser " + browser + ". Only firefox and chrome are supported.");

--- a/docker/drone/src/main/java/org/arquillian/cube/docker/drone/VideoConversionLifecycleManager.java
+++ b/docker/drone/src/main/java/org/arquillian/cube/docker/drone/VideoConversionLifecycleManager.java
@@ -2,12 +2,16 @@ package org.arquillian.cube.docker.drone;
 
 import org.arquillian.cube.docker.drone.event.AfterConversion;
 import org.arquillian.cube.spi.Cube;
+import org.arquillian.cube.spi.CubeControlException;
 import org.arquillian.cube.spi.CubeRegistry;
 import org.jboss.arquillian.core.api.Event;
 import org.jboss.arquillian.core.api.Instance;
 import org.jboss.arquillian.core.api.annotation.Inject;
 import org.jboss.arquillian.core.api.annotation.Observes;
 import org.jboss.arquillian.test.spi.event.suite.AfterSuite;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 public class VideoConversionLifecycleManager {
 
@@ -22,12 +26,16 @@ public class VideoConversionLifecycleManager {
     public void startConversion(@Observes AfterSuite afterSuite, CubeDroneConfiguration cubeDroneConfiguration,
         CubeRegistry cubeRegistry) {
 
-        if (cubeDroneConfiguration.isRecording()) {
-            initConversionCube(cubeRegistry);
-            flv2mp4.create();
-            flv2mp4.start();
-    
-            afterConversionEvent.fire(new AfterConversion());
+        try {
+            if (cubeDroneConfiguration.isRecording()) {
+                initConversionCube(cubeRegistry);
+                flv2mp4.create();
+                flv2mp4.start();
+
+                afterConversionEvent.fire(new AfterConversion());
+            }
+        } catch (CubeControlException e) {
+            Logger.getLogger(VideoConversionLifecycleManager.class.getName()).log(Level.WARNING, "Failed to start flv2mp4", e);
         }
     }
 
@@ -49,8 +57,12 @@ public class VideoConversionLifecycleManager {
     public void stopContainer(@Observes AfterConversion afterConversion) {
 
         if (this.flv2mp4 != null) {
-            flv2mp4.stop();
-            flv2mp4.destroy();
+            try {
+                flv2mp4.stop();
+                flv2mp4.destroy();
+            } catch (CubeControlException e) {
+                Logger.getLogger(VideoConversionLifecycleManager.class.getName()).log(Level.WARNING, "Failed to stop flv2mp4", e);
+            }
         }
     }
 }

--- a/docker/drone/src/main/java/org/arquillian/cube/docker/drone/VncRecorderLifecycleManager.java
+++ b/docker/drone/src/main/java/org/arquillian/cube/docker/drone/VncRecorderLifecycleManager.java
@@ -4,9 +4,13 @@ import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
 import org.arquillian.cube.docker.drone.event.AfterVideoRecorded;
 import org.arquillian.cube.docker.drone.util.VideoFileDestination;
 import org.arquillian.cube.spi.Cube;
+import org.arquillian.cube.spi.CubeControlException;
 import org.arquillian.cube.spi.CubeRegistry;
 import org.jboss.arquillian.core.api.Event;
 import org.jboss.arquillian.core.api.Instance;
@@ -34,13 +38,17 @@ public class VncRecorderLifecycleManager {
     public void startRecording(@Observes Before beforeTestMethod, CubeDroneConfiguration cubeDroneConfiguration,
         CubeRegistry cubeRegistry) {
 
-        if (cubeDroneConfiguration.isRecording()) {
+        try {
+            if (cubeDroneConfiguration.isRecording()) {
 
-            // lazy init
-            initVncCube(cubeRegistry);
+                // lazy init
+                initVncCube(cubeRegistry);
 
-            vnc.create();
-            vnc.start();
+                vnc.create();
+                vnc.start();
+            }
+        } catch (CubeControlException e) {
+            Logger.getLogger(VncRecorderLifecycleManager.class.getName()).log(Level.WARNING, "Failed to start vnc", e);
         }
     }
 
@@ -61,23 +69,27 @@ public class VncRecorderLifecycleManager {
     public void stopRecording(@Observes After afterTestMethod, TestResult testResult,
         CubeDroneConfiguration cubeDroneConfiguration, SeleniumContainers seleniumContainers) {
 
-        if (this.vnc != null) {
-            vnc.stop();
-            
-            Path finalLocation = null;
-            if (shouldRecordOnlyOnFailure(testResult, cubeDroneConfiguration)) {
-                finalLocation =
-                    moveFromVolumeFolderToBuildDirectory(afterTestMethod, cubeDroneConfiguration, seleniumContainers);
-            } else {
-                if (shouldRecordAlways(cubeDroneConfiguration)) {
+        try {
+            if (this.vnc != null) {
+                vnc.stop();
+
+                Path finalLocation = null;
+                if (shouldRecordOnlyOnFailure(testResult, cubeDroneConfiguration)) {
                     finalLocation =
                         moveFromVolumeFolderToBuildDirectory(afterTestMethod, cubeDroneConfiguration, seleniumContainers);
+                } else {
+                    if (shouldRecordAlways(cubeDroneConfiguration)) {
+                        finalLocation =
+                            moveFromVolumeFolderToBuildDirectory(afterTestMethod, cubeDroneConfiguration, seleniumContainers);
+                    }
                 }
+
+                vnc.destroy();
+
+                this.afterVideoRecordedEvent.fire(new AfterVideoRecorded(afterTestMethod, finalLocation));
             }
-
-            vnc.destroy();
-
-            this.afterVideoRecordedEvent.fire(new AfterVideoRecorded(afterTestMethod, finalLocation));
+        } catch (CubeControlException e) {
+            Logger.getLogger(VncRecorderLifecycleManager.class.getName()).log(Level.WARNING, "Failed to stop vnc", e);
         }
     }
 

--- a/docker/drone/src/main/java/org/arquillian/cube/docker/drone/util/VolumeCreator.java
+++ b/docker/drone/src/main/java/org/arquillian/cube/docker/drone/util/VolumeCreator.java
@@ -17,7 +17,7 @@ public class VolumeCreator {
 
     /**
      * Creates a new temporary folder with password file for VNC server.
-     * The folder is relative to the root of the project.
+     * The folder is relative to the root of the project target/vnc folder.
      *
      * @param password
      *     value to generate password file.
@@ -25,7 +25,7 @@ public class VolumeCreator {
      * @return Path of generated file.
      */
     public static final Path createTemporaryVolume(String password) {
-        final File tmpFile = new File(".tmp" + System.currentTimeMillis());
+        final File tmpFile = new File("target/vnc/.tmp" + System.currentTimeMillis());
 
         if (!tmpFile.mkdirs()) {
             throw new IllegalArgumentException("Temporary Folder for storing recordings could not be created.");

--- a/docker/drone/src/test/java/org/arquillian/cube/docker/drone/SeleniumContainersTest.java
+++ b/docker/drone/src/test/java/org/arquillian/cube/docker/drone/SeleniumContainersTest.java
@@ -9,6 +9,8 @@ import org.junit.runner.RunWith;
 import org.mockito.Mock;
 import org.mockito.runners.MockitoJUnitRunner;
 
+import java.util.HashMap;
+
 import static org.hamcrest.CoreMatchers.both;
 import static org.hamcrest.CoreMatchers.hasItem;
 import static org.hamcrest.CoreMatchers.is;
@@ -95,6 +97,7 @@ public class SeleniumContainersTest {
         when(cubeDroneConfiguration.isBrowserDockerfileDirectorySet()).thenReturn(false);
         when(cubeDroneConfiguration.isBrowserImageSet()).thenReturn(false);
         when(cubeDroneConfiguration.getContainerNameStrategy()).thenReturn(ContainerNameStrategy.STATIC);
+        when(cubeDroneConfiguration.getDockerRegistry()).thenReturn("");
 
         final SeleniumContainers firefox = SeleniumContainers.create("firefox", cubeDroneConfiguration);
         assertThat(firefox.getBrowser(), is("firefox"));
@@ -109,6 +112,7 @@ public class SeleniumContainersTest {
         when(cubeDroneConfiguration.isBrowserDockerfileDirectorySet()).thenReturn(false);
         when(cubeDroneConfiguration.isBrowserImageSet()).thenReturn(false);
         when(cubeDroneConfiguration.getContainerNameStrategy()).thenReturn(ContainerNameStrategy.STATIC);
+        when(cubeDroneConfiguration.getDockerRegistry()).thenReturn("");
 
         final SeleniumContainers firefox = SeleniumContainers.create("chrome", cubeDroneConfiguration);
         assertThat(firefox.getBrowser(), is("chrome"));
@@ -155,5 +159,45 @@ public class SeleniumContainersTest {
         assertThat(seleniumContainers.getSeleniumContainerName(), is("browser"));
         assertThat(seleniumContainers.getVncContainerName(), is("vnc"));
         assertThat(seleniumContainers.getVideoConverterContainerName(), is("flv2mp4"));
+    }
+
+    @Test
+    public void shouldHaveDockerRegistry() {
+
+        final CubeDroneConfiguration conf = CubeDroneConfiguration.fromMap( new HashMap<String, String>() {{
+            put("containerNameStrategy",ContainerNameStrategy.STATIC_PREFIX.toString());
+            put("dockerRegistry","my.registry");
+        }});
+
+        checkRegistry(conf);
+    }
+
+    @Test
+    public void shouldHaveDockerRegistrySlash() {
+
+        final CubeDroneConfiguration conf = CubeDroneConfiguration.fromMap( new HashMap<String, String>() {{
+            put("containerNameStrategy",ContainerNameStrategy.STATIC_PREFIX.toString());
+            put("dockerRegistry","my.registry/");
+        }});
+
+        checkRegistry(conf);
+    }
+
+    @Test
+    public void shouldHaveDockerRegistryNewline() {
+
+        final CubeDroneConfiguration conf = CubeDroneConfiguration.fromMap( new HashMap<String, String>() {{
+            put("containerNameStrategy",ContainerNameStrategy.STATIC_PREFIX.toString());
+            put("dockerRegistry","my.registry/\r\n ");
+        }});
+
+        checkRegistry(conf);
+    }
+
+    private void checkRegistry(CubeDroneConfiguration conf) {
+        final SeleniumContainers seleniumContainers = SeleniumContainers.create("firefox", conf);
+        assertThat(seleniumContainers.getSeleniumContainer().getImage().toString(), startsWith("my.registry/selenium"));
+        assertThat(seleniumContainers.getVideoConverterContainer().getImage().toString(), startsWith("my.registry/arquillian"));
+        assertThat(seleniumContainers.getVncContainer().getImage().toString(), startsWith("my.registry/richnorth"));
     }
 }

--- a/docker/ftest-container-star-operator/pom.xml
+++ b/docker/ftest-container-star-operator/pom.xml
@@ -68,6 +68,12 @@
       <artifactId>wildfly-arquillian-container-remote</artifactId>
       <version>${version.wildfly}</version>
       <scope>test</scope>
+      <exclusions>
+        <exclusion>
+          <artifactId>jconsole</artifactId>
+          <groupId>sun.jdk</groupId>
+        </exclusion>
+      </exclusions>
     </dependency>
   </dependencies>
 

--- a/docs/drone.adoc
+++ b/docs/drone.adoc
@@ -47,6 +47,7 @@ browserImage:: Docker image to be used as custom browser image instead of the of
 browserDockerfileLocation:: Dockerfile location to be used to built custom docker image instead of the official one. This property has preference over browserImage.
 containerNameStrategy:: Sets the strategy for generating the container name. Valid values are `STATIC, STATIC_PREFIX, RANDOM` with the default being `STATIC`. `STATIC` will always use the same name, `STATIC_PREFIX` lets you define a prefix and enables running different tests in parallel and `RANDOM` generates a unique container name on every test run enabling running the same test in parallel.
 containerNamePrefix:: The prefix for the `STATIC_PREFIX` container name strategy.
+dockerRegistry:: Use this to override the default docker registry with a user specified registry.
 
 IMPORTANT: Your custom images must expose the webdriver port `4444` and if you plan to use VNC, expose the default port `5900` as well
 

